pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,621 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"maps"
+	"slices"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/cilium/cilium/pkg/backoff"
+	"github.com/cilium/cilium/pkg/envoy"
+	"github.com/cilium/cilium/pkg/envoy/xds"
+
+	clusterpb "github.com/cilium/proxy/go/envoy/config/cluster/v3"
+	corepb "github.com/cilium/proxy/go/envoy/config/core/v3"
+	endpointpb "github.com/cilium/proxy/go/envoy/config/endpoint/v3"
+	listenerpb "github.com/cilium/proxy/go/envoy/config/listener/v3"
+	routepb "github.com/cilium/proxy/go/envoy/config/route/v3"
+	discoverypb "github.com/cilium/proxy/go/envoy/service/discovery/v3"
+	statuspb "google.golang.org/genproto/googleapis/rpc/status"
+)
+
+const logFeatureKey = "feature"
+
+type Interface interface {
+	Observe(ctx context.Context, typeUrl string, resourceNames []string) error
+	AddResourceWatcher(typeUrl string, cb WatcherCallback) uint64
+	RemoveResourceWatcher(id uint64)
+}
+
+var _ Interface = (*Client)(nil)
+
+type observeRequest struct {
+	typeUrl       string
+	resourceNames []string
+}
+
+type Client struct {
```

Comment:
Again, a less generic name would be good here - this will end up as `client.Client` when used at a call site from outside this package. Maybe `XDSClient`?

Reply:
Ideally it would be `xds.Client` waiting for a better name idea as `xds/client.XDSClient` doesn't sound good

Reply:
Although I agree that `xds.Client` would be ideal, that doesn't seem practical right now, so `client.XDSClient` seems like the least bad option right now.